### PR TITLE
Improve call frequency

### LINKED ISSUE #9: Save callback should not be executed while the previous wasn't resolved
I wrote such a simple sample:

```js
/**
 * @license Copyright (c) 2003-2018, CKSource - Frederico Knabben. All rights reserved.
 * For licensing, see LICENSE.md.
 */

/* globals ClassicEditor, console, window, document, setTimeout */

import { CS_CONFIG } from '@ckeditor/ckeditor5-cloud-services/tests/_utils/cloud-services-config';

ClassicEditor
	.create( document.querySelector( '#snippet-autosave' ), {
		cloudServices: CS_CONFIG,
		toolbar: {
			viewportTopOffset: 60
		},
		autosave: {
			save( editor ) {
				return saveData( editor.getData() );
			}
		}
	} )
	.then( editor => {
		window.editor = editor;
	} )
	.catch( err => {
		console.error( err.stack );
	} );

function saveData( data ) {
	return new Promise( resolve => {
		log( `Saving... (${ data })` );

		setTimeout( () => {
			log( 'Saved.' );

			resolve();
		}, 1500 );
	} );
}

function log( msg ) {
	const console = document.querySelector( '#snippet-autosave-console' );

	console.textContent = msg + '\n' + console.textContent;
}
```

```html
<div id="snippet-autosave">
	<p>TODO</p>
</div>

<pre id="snippet-autosave-console"></pre>
```

It behaves like this now:

jul-13-2018 15-29-46

IMO, it should wait with the next "Saving..." until the previous save attempt was completed.

It isn't a big deal if it doesn't wait, but it's odd. It wasn't natural for me.

### LINKED ISSUE #10: Make the throttling timeout configurable
I'm afraid that the current 500ms is too fast for most integrations.

### LINKED ISSUE #12: Save callback is triggered too often (when nothing changes)
jul-13-2018 16-12-03

The save callback is executed twice with the same content in a short period of time. I think that this is due to a flawed implementation of the `throttle()` function.

Disclaimer: I never remember which is throttling and which is debouncing, so I'll use the word throttle, but please read what's the expected semantics of it and don't focus on the name.

The throttling in autosave should work like this:

* We don't want to save characters typed one after another, so we wait until someone stopped typing for minimum X and we save the content.
* However, if someone is typing quickly for a longer period, we need to backup that in the meantime. So, if time Y passed since the first change of a series of changes, we save that anyway.

X may equal Y, but ideally they should be independent. It makes sense to make Y longer so you don't constantly save things while a person is typing. Typically, I'd set X=1s, Y=5s (or even more).

How the current implementation seems to work is this:

* First change occured – save the content.
* Don't save the content for the next X.

And the first point makes for the first "Saving... foo" while the second for another "Saving... foo".

### LINKED ISSUE #1158: core+autosave tests fail to execute together
```
npm t -- --files=core,autosave
```

Fails on a couple of plugin tests. The opposite order `autosave,core` works fine. I think that it may be due to the use of fake timers.

### LINKED ISSUE #2539: Save callback should not be executed while the previous wasn't resolved
I wrote such a simple sample:

```js
/**
 * @license Copyright (c) 2003-2018, CKSource - Frederico Knabben. All rights reserved.
 * For licensing, see LICENSE.md.
 */

/* globals ClassicEditor, console, window, document, setTimeout */

import { CS_CONFIG } from '@ckeditor/ckeditor5-cloud-services/tests/_utils/cloud-services-config';

ClassicEditor
	.create( document.querySelector( '#snippet-autosave' ), {
		cloudServices: CS_CONFIG,
		toolbar: {
			viewportTopOffset: 60
		},
		autosave: {
			save( editor ) {
				return saveData( editor.getData() );
			}
		}
	} )
	.then( editor => {
		window.editor = editor;
	} )
	.catch( err => {
		console.error( err.stack );
	} );

function saveData( data ) {
	return new Promise( resolve => {
		log( `Saving... (${ data })` );

		setTimeout( () => {
			log( 'Saved.' );

			resolve();
		}, 1500 );
	} );
}

function log( msg ) {
	const console = document.querySelector( '#snippet-autosave-console' );

	console.textContent = msg + '\n' + console.textContent;
}
```

```html
<div id="snippet-autosave">
	<p>TODO</p>
</div>

<pre id="snippet-autosave-console"></pre>
```

It behaves like this now:

jul-13-2018 15-29-46

IMO, it should wait with the next "Saving..." until the previous save attempt was completed.

It isn't a big deal if it doesn't wait, but it's odd. It wasn't natural for me.

### LINKED ISSUE #2541: Make the throttling timeout configurable
I'm afraid that the current 500ms is too fast for most integrations.

### LINKED ISSUE #2547: Save callback is triggered too often (when nothing changes)
jul-13-2018 16-12-03

The save callback is executed twice with the same content in a short period of time. I think that this is due to a flawed implementation of the `throttle()` function.

Disclaimer: I never remember which is throttling and which is debouncing, so I'll use the word throttle, but please read what's the expected semantics of it and don't focus on the name.

The throttling in autosave should work like this:

* We don't want to save characters typed one after another, so we wait until someone stopped typing for minimum X and we save the content.
* However, if someone is typing quickly for a longer period, we need to backup that in the meantime. So, if time Y passed since the first change of a series of changes, we save that anyway.

X may equal Y, but ideally they should be independent. It makes sense to make Y longer so you don't constantly save things while a person is typing. Typically, I'd set X=1s, Y=5s (or even more).

How the current implementation seems to work is this:

* First change occured – save the content.
* Don't save the content for the next X.

And the first point makes for the first "Saving... foo" while the second for another "Saving... foo".

### PR DESCRIPTION
**Pull request by ma2ciek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Other: Improved call frequency. Closes ckeditor/ckeditor5#2539. Closes ckeditor/ckeditor5#2541. Closes ckeditor/ckeditor5#2547. Closes ckeditor/ckeditor5#1158.

---

### Additional information

* This PR Introduces `Autosave#state`, which can have the following values:
	* synchronized - when all changes are saved
	* waiting - when the plugin is waiting for other changes before calling `adapter#save()` and `config.autosave.save()`
	* saving - when the provided save method is called and the plugin waits for the response from callbacks

* It switches from `throttle` to lodash's `debounce`
* It makes the timeout configurable

### CODE REVIEW

File: src/autosave.js [hunk 1]
```diff
@@ -62,6 +64,12 @@ export default class Autosave extends Plugin {
 	constructor( editor ) {
 		super( editor );
 
+		const config = editor.config.get( 'autosave' ) || {};
+
+		// A minimum amount of time that needs to pass after the last action.
+		// After that time the provided save callbacks are being called.
+		const waitingTime = config.waitingTime || 2000;
```
Comment (pjasiun) [inline src/autosave.js]: https://github.com/ckeditor/ckeditor5-autosave/issues/10#issuecomment-408118358

Comment (pjasiun) [reply src/autosave.js]: > After #12 (comment) default timeout should work in most cases and is should not need to be configurable.

Comment (pjasiun) [reply src/autosave.js]: Also, I am not sure if 2000 is not too much. Maybe 1000?

File: src/autosave.js [hunk 2]
```diff
@@ -135,13 +149,22 @@ export default class Autosave extends Plugin {
...
+				this._debouncedSave();
+			}
 
-			if ( !willOriginalFunctionBeCalled ) {
-				this._decrementCounter();
+			else if ( this.state == 'waiting' ) {
+				this._debouncedSave();
 			}
+
+			// Do nothing if the plugin is in `external-saving` state.
```
Comment (pjasiun) [inline src/autosave.js]: A comment is needed that is will be handled based on the document version.

File: src/autosave.js (hunk 1 above)
Comment (ma2ciek) [reply src/autosave.js]: > After #12 (comment) default timeout should work in most cases and it should not need to be configurable.

Right, I overlooked it.

> Also, I am not sure if 2000 is not too much. Maybe 1000?

The last call will be on the `Editor#destroy` or `Window#beforeunload`, so IMO we can use a higher value here.

**Pull request merged**